File: src/math/Vector3.ts
export class Vector3 {
    constructor(
        public x: number = 0,
        public y: number = 0,
        public z: number = 0
    ) {}
    
    set(x: number, y: number, z: number): this {
        this.x = x;
        this.y = y;
        this.z = z;
        return this;
    }

    toArray(): number[] {
        return [this.x, this.y, this.z];
    }
    
    copy(v: Vector3): this {
        this.x = v.x;
        this.y = v.y;
        this.z = v.z;
        return this;
    }
    
    add(v: Vector3): this {
        this.x += v.x;
        this.y += v.y;
        this.z += v.z;
        return this;
    }
    
    sub(v: Vector3): this {
        this.x -= v.x;
        this.y -= v.y;
        this.z -= v.z;
        return this;
    }
    
    multiplyScalar(scalar: number): this {
        this.x *= scalar;
        this.y *= scalar;
        this.z *= scalar;
        return this;
    }
    
    divideScalar(scalar: number): this {
        return this.multiplyScalar(1 / scalar);
    }
    
    length(): number {
        return Math.sqrt(this.x * this.x + this.y * this.y + this.z * this.z);
    }
    
    normalize(): this {
        return this.divideScalar(this.length() || 1);
    }
    
    dot(v: Vector3): number {
        return this.x * v.x + this.y * v.y + this.z * v.z;
    }
    
    cross(v: Vector3): Vector3 {
        return new Vector3(
            this.y * v.z - this.z * v.y,
            this.z * v.x - this.x * v.z,
            this.x * v.y - this.y * v.x
        );
    }
    
    distanceTo(v: Vector3): number {
        return Math.sqrt(
            (this.x - v.x) ** 2 +
            (this.y - v.y) ** 2 +
            (this.z - v.z) ** 2
        );
    }
}

File: iEngine-Native/Engine/src/math/Vector3.cpp
#include "iengine/math/Vector3.h"

#include <cmath>

namespace iengine {
    Vector3::Vector3(float x, float y, float z) : x(x), y(y), z(z) {}
    
    void Vector3::set(float x, float y, float z) {
        this->x = x;
        this->y = y;
        this->z = z;
    }
    
    Vector3 Vector3::copy() const {
        return Vector3(x, y, z);
    }
    
    Vector3& Vector3::add(const Vector3& v) {
        x += v.x;
        y += v.y;
        z += v.z;
        return *this;
    }
    
    Vector3& Vector3::sub(const Vector3& v) {
        x -= v.x;
        y -= v.y;
        z -= v.z;
        return *this;
    }
    
    Vector3& Vector3::multiplyScalar(float scalar) {
        x *= scalar;
        y *= scalar;
        z *= scalar;
        return *this;
    }
    
    Vector3& Vector3::divideScalar(float scalar) {
        return multiplyScalar(1.0f / scalar);
    }
    
    float Vector3::length() const {
        return std::sqrt(x * x + y * y + z * z);
    }
    
    Vector3& Vector3::normalize() {
        return divideScalar(length() > 0 ? length() : 1.0f);
    }
    
    float Vector3::dot(const Vector3& v) const {
        return x * v.x + y * v.y + z * v.z;
    }
    
    Vector3 Vector3::cross(const Vector3& v) const {
        return Vector3(
            y * v.z - z * v.y,
            z * v.x - x * v.z,
            x * v.y - y * v.x
        );
    }
    
    float Vector3::distanceTo(const Vector3& v) const {
        float dx = x - v.x;
        float dy = y - v.y;
        float dz = z - v.z;
        return std::sqrt(dx * dx + dy * dy + dz * dz);
    }
}